st2common/st2common/util/green/shell.py

```diff
@@ -165,17 +165,14 @@ def on_timeout_expired(timeout):
             # Note: We explicitly set the returncode to indicate the timeout.
             LOG.debug("Command execution timeout reached.")
 
-            # NOTE: It's important we set returncode twice - here and below to avoid race in this
-            # function because "kill_func()" is async and "process.kill()" is not.
-            process.returncode = TIMEOUT_EXIT_CODE
```

Comment:
I'm looking into this tonight.

I started by adding a test case and I was indeed able to reproduce the zombie process test issue with a local command runner like setup where we use a custom ``kill_func()``.

It seems that adding ``process.wait()`` after kill fixes the zombie process issue and we can leave that exit code assignment in place - I think that's important since if we remove it process will have a different exit code set which would break our timeout detection logic in some places.

Reply:
Also, just curious where you encountered this issue?

Was it via local shell command action runner or some other code path?

Reply:
And under which Python version you encountered this issue?

Since technically now that only support Python >= 3.6, we could also simplify that function a bit and utilize ``timeout`` parameter for ``communicate()`` method which we couldn't do then we still supported Python 2.7.

Reply:
@Kami 
> Also, just curious where you encountered this issue?

That was long-running Python actions, but next morning found zombies and started to investigate root cause of them. Finally I realized those processes just never receive kill signal, checked source of `kill`/`communicate`/`wait` and here we are :)


> It seems that adding process.wait() after kill fixes the zombie process issue and we can leave that exit code assignment in place - I think that's important since if we remove it process will have a different exit code set which would break our timeout detection logic in some places.

No, `wait()` won't work in both cases (eventlet and gevent) because this method checks `returncode` against `None` to proceed, otherwise it will just return existing `returncode`. Also it should not cause any mess with timeout detection logic becuse it will be set to `-9` right after `wait()`.


> And under which Python version you encountered this issue?

Python 3.6. I'll be able to look into using `communicate()` here tonight.

Reply:
Using ``communicate()`` is actually not a blocker, just a potential future simplification :)

And thanks for answering those questions.

I think the change which was merged should cover all those scenarios and I believe we now also have integration tests for most of those situations (shell=True as is the case for local runner, shell=False for python action runner, etc. Only thing we are missing is tests for read_stdout_func + read_stdout_func which verify threads which call those functions are correctly stopped / killed on timeout.).

Can you please confirm the change which has been merged indeed fixes the issue (I believe it should since it's based on your changes)?

Reply:
> Using communicate() is actually not a blocker, just a potential future simplification :)
I see :)

> Can you please confirm the change which has been merged indeed fixes the issue (I believe it should since it's based on your changes)?
Just tried to reproduce the issue and everything seem to be fine:
1. subprocess is getting killed on timeout
2. it lefts no zombie process after kill process
3. sure, subprocess might be tracked down within ST2 as timed out process

Thank you for helping with my PR

Reply:
Great. Thanks for confirming and submitting this bug fix.